### DISCUSSION TITLE
Non-Player Creatures that use player avatar animation combinations (male/female armour + weapons)?

### DISCUSSION THREAD

**Discussion by WithinAmnesia** [flareteam/flare-game#867] [Q&A] [answered]
Can the Flare Engine do what most large scale R.P.G.s do where they save work and add diversity of content by having N.P.C.s. (friendly / neutral / enemy) have the creative options to render their avatars / sprites such as the same appearance as the player's avatar / sprites in terms of possible armour / weapon male/female sprite ~'load out'?

Pre-baked creature animation assets are different from the system used for the player avatar animation combinations (male/female armour + weapons). These pre-baked creature animation assets are very limited in scope and offer much less creative power in terms of amount of creative work input verses game content output.

It with be incredibly creatively empowering if the Flare Engine has the possibility to support / have a ~'creative tool' / harness a feature to implement Non-Player Creatures that use player avatar animation combinations (male/female armour + weapons). For this could lead to hundreds of thousands of different possible N.P.C. appearance option combinations to drive large scale game content generation; while also using the same graphical sprites as what the player avatar has available to render / display with. In effect this is killing two birds with one stone in recycling graphical avatar animation resources and also adding tremendous content depth and developer creative power.

Examples of official release R.P.G.s with Non-Player Creatures that use player avatar animation combinations (male/female armour + weapons):

Baldur's Gate (1998)
Everquest (1999)
RuneScape (2001)
World of Warcraft (2004)
Dragon Age (2009)
Path of Exile (2013)

There are more examples but those are enough. This is a classic developer work saving, content generating feature that I hope the Flare Engine can pull off and support. What is one the line here is literally the difference of a small scale game and a massive scale game in terms of content generation possibility for the same amount of developer work.

**Comment by dorkster** [comment, marked as answer]
The way that player animation layers are handled is very closely tied to the inventory menu. So if we decide to implement this, one of two paths should be taken:

1. Decouple the inventory behavior from the menu and let every entity have an inventory.
2. Add a way to define additional animation layers to the non-player entity StatBlocks.

Number 1 is probably the "right" option, but I think I might favor number 2 due to simplicity.

**Reply by WithinAmnesia** [reply]
Would 'Option 1' possibly lead to a 'neutral' NPC system where there are effectively non-player characters that use A.I. but use a player inventory / 'character stat sheet' (just as with Dungeon and Dragons / Baldur's Gate)?

An A.I. controlled player 'rig' / creature where the creature / NPC can have attributes / items / abilities / powers like a player character yet with A.I. control scripts to control the actions of this 'None-Player Character; can this be done in Flare?

'Option 1' could lead to NPCs being able to show items they are equipped with and they could drop said equipped items on death (as supported by Baldur's Gate and RuneScape). This system / 'Option 1' could support a system such as with Dungeons and Dragons / Baldur's Gate where the player can see what a humanoid character is equipped with and make better informed decisions on where they kill or avoid / negotiate. Also this is like with WoW and Runescape where a player can judge another humanoid based on what they are wearing (but this is usually only other players, but still cool).

I remember a Legend in Everquest player history where since mobs / killable creatures would display what they were equipped with a player could see for example if a monster would drop a weapon they were wielding. There was a rare case where a unique and 'unkillable' mobs would wield a legendarily power sword and it was a goal of the mightiest players to find a way to kill this ~'unkillable' mob to get a legendary reward. Against the odds the strongest guild in Everquest killed the unkillable mob and a Paladin wielded the legendary sword unthinkable by the devs the player could obtain; but they did. The sword was called Inte's Sword of Death and the player was called Kreugen and the guild was called Fires of Heaven.

Screen shot: https://web.archive.org/web/20030823015708/http://www.everlore.com/files2/FLO148346.jpg
"The paly sword in air was Inte's Sword of Death it only dropped once in game and that was for FoH. Kreugen got it at the time there were no NO RENT tags and it poofed after he zoned. Here is a pic of him wielding it. and if you guys were around back that you'll know that those kind of numbers were godly.
<img src="http://www.everlore.com/files2/FLO148346.jpg"/>
Stats were 15 dly, 150 dmg
15 str +12 wis +12 mana +25 SV cold/magic/fire/dis/poison +5
Effect: Flame Shock

It was put on him as a joke Alla Seru's sword (actually this is prolly why Seru's sword was made none/none. Up until PoP they couldn't add weapons as part of the model **so they actually had to equip it as a weapon**. They though no one would ever kill Inte so they put the FA model on him with these stats"
http://thedruidsgrove.org/archive/eq/t-41.html
https://www.firesofheaven.org/threads/everquest-mysteries.471/page-27

I would like to have 'Option 1' be supported even if it takes longer. For I feel such support for None-Player Characters with their own inventory and displaying what they are equipped with can be a system that can support past legendary adventure tales. Such as with Kreugen and the legendary Inte's Sword of Death; I feel that is something worth supporting. Also 'Option 1' system can support this intriguing kind of optional player 'holy grail / treasure / legendary item hunt' content as found in Path of Exile: The Oni Goroshi of the secret legendary from Uber Hillock (a very rare 'Uber' version of the first boss) in Path of Exile: https://pathofexile.fandom.com/wiki/Oni-Goroshi 

What challenges does 'Option 1' face to implement? It has a ton of benefits, although I am also a bit in the dark as to the scope of difficulty in order to realize it. Yet if 'Option 1' can be realized it would make Flare on par with Baldur's Gate, RuneScape and elements of Everquest and World of Warcraft (with an honourable mention to Path of Exile) in terms of player item hunting / Player loot progression. Of which is an important core driving element to these kind of loot driven R.P.G.s such as the Diablo series.

Also it would be just really cool to be able to support the 'soul' of past legends and mysteries of old in a new light.

**Reply by WithinAmnesia** [reply]
Also possibly 1.14?

**Reply by dorkster** [reply]
This would take a lot of work, and I don't see it as critical to the type of ARPG Flare is. As it is, no target version is currently planned.

However, option 1 would help tackle some other issues. For example, it would be nice if the Goblin Thief enemy had an inventory to store its stolen potions (and thus be able to drop them on death). So maybe some day. Just don't expect it any time soon.

**Reply by WithinAmnesia** [reply]
1.20?XD

**Reply by dorkster** [reply]
That'd be more realistic.

**Reply by WithinAmnesia** [reply]
@dorkster Wait how simple / complex / difficult to implement is Option 2? Can these solution options be broken into different parts for different patches? Possibly 1.14 for Option 2 and ~'1.XX' for Option 2? As in like adding Option 2 animation layers for humanoid non-player entity support first with a later possibility of adding Option 1 entities with inventories?

I am thinking of minimum viable solution with 'enemy looks like player and can be customized like a player' without going too potentially complex. Is this a case where player animation graphics / assets could be added as an option for non-player entities; and manually tweaked to say have a humanoid non-player entity look like a player and have armour and have visual weapon combinations as a player character? Yet behind the scenes these are manually tweaked visual details of a developer / advanced creature / entity sprite / graphics display? Thus if a developer wants to simulate displayed / worn items dropped on death from a non-player entity they could just add a set of 100% drop rate items; to kind of simulate a more advanced inventory system but use a simpler graphical solution of using player sprite armour and weapon layer combinations manually for each entity tweaked behind the scenes?

Can Option 2 be done much more easily than Option 1 and possibly can Option 1 be added later (if at all, if it turns out that Option 1 is just too complex)? Is this a case of where Option 1 is around ~8/10 difficult and Option 2 is ~3/10 difficult (or something like that)? I am not sure the severity of the difference of difficulty to implement Option 1 verses Option 2. Or if more pragmatically they could both be mutual options done / eventually implemented in sequence; as Option 2 possibly followed by Option 1? I am wondering if Option 2 could possibly be a reasonable way forward as a minimum viable solution for humanoid non-player entities that can use player avatar animation combinations (male/female armour + weapons)?

**Reply by dorkster** [reply]
Option 2 would probably be one step of implementing Option 1, so it would naturally be easier to implement just that step. In fact, I'll cautiously say it might be simple enough to squeeze in for 1.13. I've opened an issue for it here: https://github.com/flareteam/flare-engine/issues/1802
